firedrake/output.py

```diff
@@ -651,3 +756,366 @@ def write(self, *functions, **kwargs):
+                                   '/topology/order',
+                                   '/topology/orientation']):
+        return ViewerHDF5.Format.HDF5_PETSC
+    elif all(d in h5pyfile for d in ['/labels/celltype',
+                                     '/geometry/vertices',
+                                     '/viz/topology/cells']):
+        return ViewerHDF5.Format.HDF5_XDMF
+    else:
+        raise RuntimeError("Unknown HDF5 file format used in %s", viewer.getFileName())
+    del h5pyfile
+
+
+def _dmload_plex(viewer, name):
+    plex = PETSc.DMPlex()
+    plex.create(comm=viewer.comm)
+    plex.setName(name)
+    format = _get_format(viewer)
+    viewer.pushFormat(format=format)
+    if format == ViewerHDF5.Format.HDF5_PETSC:
+        sfXB = plex.topologyLoad(viewer)
+        plex.coordinatesLoad(viewer)
+        plex.labelsLoad(viewer)
+        plex.setOptionsPrefix("loaded_")
+        plex.viewFromOptions("-dm_view")
+    elif format == ViewerHDF5.Format.HDF5_XDMF:
+        plex.load(viewer=viewer)  # TODO: load labels, too.
+        plex.setOptionsPrefix("loaded_")
+        plex.viewFromOptions("-dm_view")
+        plex.interpolate()
+        plex.setOptionsPrefix("interpolated_")
+        plex.viewFromOptions("-dm_view")
+        sfXB = None
+    viewer.popFormat()
+    return plex, sfXB
+
+
+from firedrake.functionspacedata import get_global_numbering, get_shared_data_key
+def get_shared_data_key_for_checkpointing(mesh, element):
+    flag = isinstance(element, (ufl.VectorElement, ufl.TensorElement))
+    return get_shared_data_key(mesh, element, use_true_tensor_element_shape=flag)
+
+
+def get_sectiondm_name(nodes_per_entity, real_tensorproduct):
+    return "_".join(["sectiondm"] +
+                    [str(n) for n in nodes_per_entity] +
+                    [str(real_tensorproduct)])
+
+
+def get_sectiondm_for_checkpointing(tV):
+    sd_key = get_shared_data_key_for_checkpointing(tV.mesh(), tV.ufl_element())
+    dm = tV.mesh().topology_dm
+    if isinstance(tV.ufl_element(), (ufl.VectorElement, ufl.TensorElement)):
+        nodes_per_entity, real_tensorproduct = sd_key
+        assert not real_tensorproduct
+        global_numbering = get_global_numbering(tV.mesh(), sd_key)
+        sectiondm = PETSc.DMShell().create(dm.comm)
+        sectiondm.setPointSF(dm.getPointSF())
+        sectiondm.setSection(global_numbering)
+    else:
+        sectiondm = tV._shared_data.node_set.halo.dm
+    sectiondm.setName(get_sectiondm_name(*sd_key))
+    return sectiondm
+
+
+class Dict(dict): pass
+
+
+class CheckpointFile(object):
+    _mesh_cache = weakref.WeakValueDictionary()
+
+    def __init__(self, filename, mode, comm=COMM_WORLD):
+        self.viewer = ViewerHDF5()
+        self.viewer.create(filename, mode=mode, comm=comm)
+        commkey = comm.py2f()
+        assert commkey != MPI.COMM_NULL.py2f()
+        self._meshes = self._mesh_cache.setdefault((filename, commkey), Dict())
```

Comment:
I don't think this does what you think it does:
```
from weakref import WeakValueDictionary as wvd
import gc

foo = wvd()

class Dict(dict):
    pass

baz = object()
bar = Dict()
bar[1] = baz
foo[2] = bar
del bar
gc.collect()
print(list(foo.keys()))
=> []
```

Whereas I think you're trying to hold on to the meshes you've loaded as long as they are in use?

Reply:
Yes. Thanks. I tried to fix it. I hope it is fine now.